[Export] Improve metadata and output parsing during deserialization

jwfromm opened the pull request:
Summary:
Deserialization of metadata could encounter a bug where commas are used in valid metadata names. This specifically occurs when a split of a `torch.nn.Sequential` stack is used, but may have other possible triggers. Because the deserialization relies on a comma based string split, such names trigger an error. This change uses a simple regular expression to ignore commas within parentheses to avoid the issue.

I add a test that constructs one such problematic sequential stack and show that it can be properly round-tripped with the improved splitting.

Similarly, deserialization could fail when outputs are not a tensor type. Although such outputs like None or constants are not very useful, they do show up in graphs and export should be able to support them. This change improves output node parsing and adds a corresponding test.

Test Plan: buck test //caffe2/test:test_export -- TestSerialize

Differential Revision: D55391674

jwfromm commented:
@zhxchen17, I think this PR is in pretty good shape, could you give it a look?

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -1516,7 +1521,7 @@ def deserialize_graph(self, serialized_graph: Graph) -> torch.fx.Graph:
         else:
             output_node.meta["val"] = tuple(
-                arg.meta["val"] for arg in output_node.args[0]
+                arg.meta["val"] if hasattr(arg, "meta") else arg for arg in output_node.args[0]
```
```suggestion
                arg.meta["val"] if isinstance(arg, Node) else arg for arg in output_node.args[0]
```

zhxchen17 commented on torch/_export/serde/serialize.py:
```diff
@@ -1976,8 +1981,16 @@ def deserialize_meta_func(serialized_target: str):
+            # encapsulated by parens, which are valid traces.
+            def metadata_split(metadata):
+                # Remove the parentheses and commas inside them
+                metadata = re.sub(r'\(.*?\)', '', metadata)
```
@jwfromm I probably forgot about the edge cases, could you give one example of what this looks like?

jwfromm replied:
Yeah its a weird case. Specifically when you use a subsection of a Sequential stack of layers as in this module
```
class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.layers = torch.nn.Sequential(
            torch.nn.SiLU(),
            torch.nn.SiLU(),
            torch.nn.SiLU(),
        )

    def forward(self, x):
        # Splitting layers of a sequential stack introduces commas and parens
        # into metadata trace.
        out_start, out_rest = self.layers[0], self.layers[1:]
        h = out_start(x)
        h = out_rest(h)
        return h
  ```
  
  You end up with metadata in the nodes that looks like this
  ```
  {'stack_trace': '  File "/Users/jwfromm/Documents/Scripts/test_compile.py", line 18, in forward\n    h = out_rest(h)\n  File "/Users/jwfromm/Library/Python/3.9/lib/python/site-packages/torch/nn/modules/module.py", line 1520, in _call_impl\n    return forward_call(*args, **kwargs)\n  File "/Users/jwfromm/Library/Python/3.9/lib/python/site-packages/torch/nn/modules/container.py", line 217, in forward\n    input = module(input)\n', 'nn_module_stack': {'L__self__': ('', <class '__main__.MyModule'>), '5016173536': ('layers.slice(1, None, None)', <class 'torch.nn.modules.container.Sequential'>), 'sub2': ('layers.slice(1, None, None).1', <class 'torch.nn.modules.activation.SiLU'>)}, 'source_fn_stack': [('sub2', <class 'torch.nn.modules.activation.SiLU'>)], 'original_aten': <OpOverload(op='aten.silu', overload='default')>, 'from_node': [('h_1', 'sub2')], 'seq_nr': -1, 'val': FakeTensor(..., size=(1, 3, 32, 32)), 'tensor_meta': TensorMetadata(shape=torch.Size([1, 3, 32, 32]), dtype=torch.float32, requires_grad=False, stride=(3072, 1024, 32, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})}
 ```
 
 The important bit being the bits like `'layers.slice(1, None, None).1',`. When you slice the sequential stack, it shows up in the trace as a `.slice(idx, None, None)`. The extra commas in there break the metadata parsing. I think there's also a good argument that it should be represented with a more clear name but that's a much more involved fix.

zhxchen17 replied:
I see you mean now. Thanks for clarification.

zhxchen17 replied:
@jwfromm could you add a TODO here to remind us of a more involved fix in the future?

zhxchen17 reviewed (commented):
thanks! a small nit and a question

jwfromm committed:
[Export] Improve metadata and output parsing during deserialization (#122793)

Summary:

Deserialization of metadata could encounter a bug where commas are used in valid metadata names. This specifically occurs when a split of a `torch.nn.Sequential` stack is used, but may have other possible triggers. Because the deserialization relies on a comma based string split, such names trigger an error. This change uses a simple regular expression to ignore commas within parentheses to avoid the issue.

I add a test that constructs one such problematic sequential stack and show that it can be properly round-tripped with the improved splitting.

Similarly, deserialization could fail when outputs are not a tensor type. Although such outputs like None or constants are not very useful, they do show up in graphs and `export` should be able to support them. This change improves output node parsing and adds a corresponding test.

Test Plan: buck test //caffe2/test:test_export -- TestSerialize

Differential Revision: D55391674

zhxchen17 reviewed (approved):
LGTM! Thanks for contributing :)

jwfromm commented:
@pytorchmergebot merge
